### ISSUE TITLE
I am trying to train the FastConformer 120M model from scratch, but it is not converging?

### ISSUE THREAD

**Issue by PhamDangNguyen** [NVIDIA/NeMo#12167]
I am trying to train the FastConformer 120M model from scratch, but it is not converging. This is my config base:
- GPU: 40G
- Batch_size = 32
- amount of data: 5000 hours (4.000.000 data points).
'''
name: "FastConformer-CTC-BPE"

model:
  sample_rate: 16000
  log_prediction: true # enables logging sample predictions in the output during training
  ctc_reduction: 'mean_volume'
  skip_nan_grad: false

  train_ds:
    manifest_filepath: /home/team_voice/STT_pdnguyen/finetune-fast-conformer/metadata_train/train_remove_func_18.json
    sample_rate: ${model.sample_rate}
    batch_size: 40 # you may increase batch_size if your memory allows
    shuffle: true
    num_workers: 32
    pin_memory: true
    max_duration: 21 # it is set for LibriSpeech, you may need to update it for your dataset
    min_duration: 0.3
    # tarred datasets
    is_tarred: false
    tarred_audio_filepaths: null
    shuffle_n: 2048
    # bucketing params
    bucketing_strategy: "fully_randomized"
    bucketing_batch_size: null

  validation_ds:
    manifest_filepath: /home/team_voice/STT_pdnguyen/finetune-fast-conformer/metadata_train/test_remove_func.json
    sample_rate: ${model.sample_rate}
    batch_size: 16 # you may increase batch_size if your memory allows
    shuffle: false
    use_start_end_token: false
    num_workers: 16
    pin_memory: true

  test_ds:
    manifest_filepath: /home/team_voice/STT_pdnguyen/finetune-fast-conformer/metadata_train/test_remove_func.json
    sample_rate: ${model.sample_rate}
    batch_size: 16 # you may increase batch_size if your memory allows
    shuffle: false
    use_start_end_token: false
    num_workers: 16
    pin_memory: true

  tokenizer:
    dir: /home/team_voice/STT_pdnguyen/finetune-fast-conformer/dict_N/tokenizer_spe_bpe_v3072  # path to directory which contains either tokenizer.model (bpe) or vocab.txt (wpe)
    type: bpe  # Can be either bpe (SentencePiece tokenizer) or wpe (WordPiece tokenizer)

  preprocessor:
    _target_: nemo.collections.asr.modules.AudioToMelSpectrogramPreprocessor
    sample_rate: 16000
    normalize: "per_feature"
    window_size: 0.025
    window_stride: 0.01
    window: "hann"
    features: 80
    n_fft: 512
    log: true
    frame_splicing: 1
    dither: 0.00001
    pad_to: 0
    pad_value: 0.0

  spec_augment:
    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
    freq_masks: 2 # set to zero to disable it
    time_masks: 0 
    freq_width: 27
    time_width: 0.05

  encoder:
    _target_: nemo.collections.asr.modules.ConformerEncoder
    feat_in: ${model.preprocessor.features}
    feat_out: -1 # you may set it if you need different output size other than the default d_model
    n_layers: 19
    d_model: 512

    subsampling: dw_striding # vggnet, striding, stacking or stacking_norm, dw_striding
    subsampling_factor: 8 # must be power of 2 for striding and vggnet
    subsampling_conv_channels: 256 # -1 sets it to d_model
    causal_downsampling: false

    ff_expansion_factor: 4

    self_attention_model: rel_pos # rel_pos or abs_pos
    n_heads: 8 # may need to be lower for smaller d_models
    att_context_size: [-1, -1] # -1 means unlimited context
    att_context_style: regular # regular or chunked_limited
    xscaling: true # scales up the input embeddings by sqrt(d_model)
    untie_biases: true # unties the biases of the TransformerXL layers
    pos_emb_max_len: 5000

    conv_kernel_size: 9
    conv_norm_type: 'batch_norm' # batch_norm or layer_norm or groupnormN (N specifies the number of groups)
 conv_context_size[0]+conv_context_size[1]+1==conv_kernel_size
    conv_context_size: null

    dropout: 0.1 # The dropout used in most of the Conformer Modules
    dropout_pre_encoder: 0.1 # The dropout used before the encoder
    dropout_emb: 0.0 # The dropout used for embeddings
    dropout_att: 0.1 # The dropout for multi-headed attention modules

    stochastic_depth_drop_prob: 0.0
    stochastic_depth_mode: linear  # linear or uniform
    stochastic_depth_start_layer: 1

  decoder:
    _target_: nemo.collections.asr.modules.ConvASRDecoder
    feat_in: null
    num_classes: -1
    vocabulary: []

  interctc:
    loss_weights: []
    apply_at_layers: []

  optim:
    name: adamw
    lr: 1e-3
    betas: [0.9, 0.98]
    weight_decay: 1e-3

    sched:
      name: CosineAnnealing
      warmup_ratio: 0.1
      min_lr: 5e-4

trainer:
  devices: 3 # number of GPUs, -1 would use all available GPUs
  num_nodes: 1
  max_epochs: 1000
  max_steps: -1 # computed at runtime if not set
  val_check_interval: 1.0 # Set to 0.25 to check 4 times per epoch, or an int for number of iterations
  accelerator: auto
  strategy: ddp
  accumulate_grad_batches: 1
  gradient_clip_val: 0.0
  precision: 32 # 16, 32, or bf16
  log_every_n_steps: 10  # Interval of logging.
  enable_progress_bar: True
  num_sanity_val_steps: 0 # number of steps to perform validation steps for sanity check the validation process before starting the training, setting to 0 disables it
  check_val_every_n_epoch: 1 # number of evaluations on validation every n epochs
  sync_batchnorm: true
  enable_checkpointing: False  # Provided by exp_manager
  logger: false  # Provided by exp_manager
  benchmark: false # needs to be false for models with variable-length speech input as it slows down training

exp_manager:
  exp_dir: null
  name: ${name}
  create_tensorboard_logger: true
  create_checkpoint_callback: true
  checkpoint_callback_params:
    monitor: "val_wer"
    mode: "min"
    save_top_k: 40
    always_save_nemo: True # saves the checkpoints as nemo files instead of PTL checkpoints

  resume_from_checkpoint: /home/team_voice/STT_pdnguyen/Backup_version_train_L40S/fast_conformer_ja_larg/25_12_2/nemo_experiments/FastConformer-CTC-BPE/2025-02-11_08-17-59/checkpoints/FastConformer-CTC-BPE--val_wer=1.0000-epoch=3.ckpt
  resume_if_exists: true
  resume_ignore_no_checkpoint: true

  create_wandb_logger: false
  wandb_logger_kwargs:
    name: null
    project: null
'''

How can I change my config to make it converge? Thanks for any help!

**Comment by nithinraok** [comment]
Hi, sorry for late response. Is it possible for you share the plots on the current convergence.

**Comment by PhamDangNguyen** [comment]
> Hi, sorry for late response. Is it possible for you share the plots on the current convergence.
Of course, this is my WER convergence at each step, quite bad.
![Image](https://github.com/user-attachments/assets/a1ac9e63-5764-41b4-ab4f-9d785f4d814c)

**Comment by nithinraok** [comment]
Try increasing `accumulate_grad_batches` to 4 and try different values to indirectly increase your effective batch_size. I cannot suggest more without knowing your dataset size and current loss curves. 

Try with a larger vocab size as well. (like 256 or 512)

**Comment by PhamDangNguyen** [comment]
> Try increasing `accumulate_grad_batches` to 4 and try different values to indirectly increase your effective batch_size. I cannot suggest more without knowing your dataset size and current loss curves.
> 
> Try with a larger vocab size as well. (like 256 or 512)

Sorry, I will provide more information about the training process below:
My dataset is a Japanese dataset. To preprocess the text, I am using the notebook https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/ASR_CTC_Language_Finetuning.ipynb

then, when i have created  my vocab by process_asr_text_tokenizer.py by nemo code, this is my setup 
python  process_asr_text_tokenizer.py \
        --manifest="/home/team_voice/STT_pdnguyen/finetune-fast-conformer/metadata_train/raw_train.json" \
        --data_root="/home/team_voice/STT_pdnguyen/finetune-fast-conformer/dict_N" \
        --vocab_size=3072\
        --tokenizer="spe" \
        --no_lower_case \
        --spe_type="bpe" \
        --spe_character_coverage=1.0 \
        --log

However, I got a minimum of 4,588 tokens > 3,072 tokens I set up (that is an error). After removing 18 least frequent tokens in my ground truth text, I finally obtained a vocabulary and tokenizer of size 3,072.
Here are my training results:

![Image](https://github.com/user-attachments/assets/b5658c42-c891-4147-85ad-2cee8ac58929)

![Image](https://github.com/user-attachments/assets/1f1339b7-d0d8-4274-8dbb-2f2f51c52bb0)

![Image](https://github.com/user-attachments/assets/48303d84-c8b3-4a4d-93ed-d70c07905435)

![Image](https://github.com/user-attachments/assets/3533c402-bcc7-4d1a-8be3-cf6932dca51f)

![Image](https://github.com/user-attachments/assets/b584e53b-8692-4c93-b2f0-de3b5dd9f419)

**Comment by nithinraok** [comment]
From recent experiments with japanese language we noticed we need quite large vocab size. @ipmedenn any suggestions?

**Comment by PhamDangNguyen** [comment]
> From recent experiments with japanese language we noticed we need quite large vocab size. [@ipmedenn](https://github.com/ipmedenn) any suggestions?

yes, for Japanese i make size of vocab is 3072. Because i refer by this repo https://huggingface.co/nvidia/parakeet-tdt_ctc-0.6b-ja

**Comment by ipmedenn** [comment]
Vocab size of 3072 is good for Japanese, I don't think you need to increase it further.

**Comment by ipmedenn** [comment]
As I can see, your train_loss does not decrease at all, which could mean some issue in training data. I recommend to recognize a subset of train with https://huggingface.co/nvidia/parakeet-tdt_ctc-0.6b-ja model, and look at CER.
Also, to have CER instead of WER in validation, make sure that you set model.use_cer=True in config.

**Comment by PhamDangNguyen** [comment]
> As I can see, your train_loss does not decrease at all, which could mean some issue in training data. I recommend to recognize a subset of train with https://huggingface.co/nvidia/parakeet-tdt_ctc-0.6b-ja model, and look at CER. Also, to have CER instead of WER in validation, make sure that you set model.use_cer=True in config.

How can i set CER instead WER with Faseconformer's [config](http://github.com/NVIDIA/NeMo/blob/main/examples/asr/asr_ctc/speech_to_text_ctc_bpe.py)? 
I using ReazonSpeech v2.0 which same the [parakeet-tdt_ctc-0.6b-ja](https://huggingface.co/nvidia/parakeet-tdt_ctc-0.6b-ja) training and JSUT for validate data. Using NeMo's processing template for Japanese data, I still had to remove characters appearing fewer than 18 times in the training set to achieve a vocabulary size of 3072. Could this make the model harder to converge?

**Comment by PhamDangNguyen** [comment]
.
